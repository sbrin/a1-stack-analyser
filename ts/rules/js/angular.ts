import { register } from '../../register.js';

register({
  tech: 'angular',
  name: 'Angular',
  type: 'language',
  files: ['angular.json'],
  dependencies: [
    { type: 'npm', name: 'angular' },
    { type: 'npm', name: /^@angular\//, example: '@angular/forms' },
  ],
});
