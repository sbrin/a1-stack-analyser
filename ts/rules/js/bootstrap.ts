import { register } from '../../register.js';

register({
  tech: 'bootstrap',
  name: 'Bootstrap',
  type: 'tool',
  dependencies: [{ type: 'npm', name: 'bootstrap' }],
});
