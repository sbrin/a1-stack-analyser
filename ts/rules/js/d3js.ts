import { register } from '../../register.js';

register({
  tech: 'd3js',
  name: 'D3.js',
  type: 'tool',
  dependencies: [
    { type: 'npm', name: 'd3' },
    { type: 'npm', name: 'd3-shape' },
    { type: 'npm', name: '@types/d3' },
    { type: 'npm', name: '@types/d3-shape' },
  ],
});
