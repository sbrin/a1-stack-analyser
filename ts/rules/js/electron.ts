import { register } from '../../register.js';

register({
  tech: 'electron',
  name: 'Electron',
  type: 'tool',
  dependencies: [
    { type: 'npm', name: 'electron' },
    { type: 'githubAction', name: 'samuelmeuli/action-electron-builder' },
  ],
});
