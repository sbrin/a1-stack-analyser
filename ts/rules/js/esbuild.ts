import { register } from '../../register.js';

register({
  tech: 'esbuild',
  name: 'Esbuild',
  type: 'tool',
  dependencies: [{ type: 'npm', name: 'esbuild' }],
});
