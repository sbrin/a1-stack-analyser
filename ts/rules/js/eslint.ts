import { register } from '../../register.js';

register({
  tech: 'eslint',
  name: 'Eslint',
  type: 'tool',
  files: ['.eslintrc', '.eslintrc.cjs', '.eslintrc.json', '.eslintrc.js'],
  dependencies: [{ type: 'npm', name: 'eslint' }],
});
