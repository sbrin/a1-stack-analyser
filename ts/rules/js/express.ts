import { register } from '../../register.js';

register({
  tech: 'express',
  name: 'Express',
  type: 'tool',
  dependencies: [{ type: 'npm', name: 'express' }],
});
