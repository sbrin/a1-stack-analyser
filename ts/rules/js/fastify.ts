import { register } from '../../register.js';

register({
  tech: 'fastify',
  name: 'Fastify',
  type: 'tool',
  dependencies: [
    { type: 'npm', name: 'fastify' },
    { type: 'npm', name: '@opentelemetry/instrumentation-fastify' },
  ],
});
