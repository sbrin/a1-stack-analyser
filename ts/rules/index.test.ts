import { describe, expect, it } from 'vitest';

import { dependencies, rawList } from '../loader.js';
import { registeredRules, registeredTech } from '../register.js';

import '../autoload.js';

describe('all', () => {
  it('should match everything', async () => {
    expect(Array.from(registeredTech.values()).sort()).toMatchSnapshot();
  });

  it('should not list multiple times the same name', () => {
    const names = new Set<string>();
    for (const rule of registeredRules) {
      if (names.has(rule.name)) {
        expect(rule.name).toBe(false);
        continue;
      }

      names.add(rule.name);
    }
  });

  it('should not list multiple times the same dependencies', () => {
    const deps = new Set<string>();
    for (const rule of registeredRules) {
      if (!rule.dependencies) {
        continue;
      }

      for (const dep of rule.dependencies) {
        const str = `${dep.type}||${dep.name}`;
        if (deps.has(str)) {
          expect(str).toBe(false);
          continue;
        }

        deps.add(str);
      }
    }
  });

  it('should not multiple tech', () => {
    for (const [name, matchers] of Object.entries(dependencies)) {
      const examples = new Set<string>();

      for (const item of rawList) {
        if (item.type !== 'dependency' || item.ref.type !== name) {
          continue;
        }

        const ex = 'example' in item.ref ? item.ref.example : item.ref.name;
        if (examples.has(ex)) {
          expect(ex).toBe(false);
          continue;
        }

        examples.add(ex);
      }

      for (const ex of examples) {
        let matched: string | null = null;
        for (const matcher of matchers) {
          if (!matcher.match.test(ex)) {
            continue;
          }
          if (matched) {
            if (matched.split('.')[0] === matcher.tech.split('.')[0]) {
              // Skip hosting
              continue;
            }
            throw new Error(
              `Dependency "${ex}" (${name}) was already matched by "${matched}", and again by "${matcher.tech}"`
            );
          }
          matched = matcher.tech;
        }
      }
    }
  });
});
