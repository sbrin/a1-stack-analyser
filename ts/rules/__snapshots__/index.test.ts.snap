// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`all > should match everything 1`] = `
[
  "adminer",
  "adobe",
  "adobecommercecloud",
  "adyen",
  "airbyte",
  "airtable",
  "algolia",
  "alibabacloud",
  "alpinejs",
  "amplitude",
  "angular",
  "ansible",
  "apache_airflow",
  "apache_cassandra",
  "apache_couchdb",
  "apache_flink",
  "apache_hadoop",
  "apache_hive",
  "apache_iceberg",
  "apache_kafka",
  "apache_solr",
  "apache_spark",
  "apache_storm",
  "apache_thrift",
  "apiplatform",
  "appveyor",
  "assemble",
  "astro",
  "atlasgo",
  "atlassian",
  "atlassian.bitbucket",
  "atlassian.confluence",
  "atlassian.jira",
  "atlassian.opsgenie",
  "atlassian.trello",
  "auth0",
  "aws",
  "aws.amplifyhosting",
  "aws.apigateway",
  "aws.athena",
  "aws.cloudformation",
  "aws.cloudfront",
  "aws.cloudsearch",
  "aws.cloudwatch",
  "aws.codebuild",
  "aws.documentdb",
  "aws.dynamodb",
  "aws.ebs",
  "aws.ec2",
  "aws.ecr",
  "aws.ecs",
  "aws.efs",
  "aws.eks",
  "aws.elasticache",
  "aws.fargate",
  "aws.glacier",
  "aws.glue",
  "aws.kafka",
  "aws.kinesis",
  "aws.kms",
  "aws.lambda",
  "aws.lightsail",
  "aws.memorydb",
  "aws.mq",
  "aws.neptune",
  "aws.opensearch",
  "aws.polly",
  "aws.rds",
  "aws.redshift",
  "aws.rekognition",
  "aws.s3",
  "aws.sagemaker",
  "aws.secretsmanager",
  "aws.ses",
  "aws.sfn",
  "aws.sns",
  "aws.sqs",
  "aws.timestream",
  "aws.translate",
  "azure",
  "azure.aks",
  "azure.ci",
  "azure.cosmosdb",
  "azure.functions",
  "azure.mariadb",
  "azure.mysql",
  "azure.postgres",
  "azure.redis",
  "azure.sql",
  "azure.staticwebapps",
  "azure.storage",
  "bash",
  "betterstack",
  "bigcommerce",
  "blackfire",
  "blitzjs",
  "bootstrap",
  "box",
  "brevo",
  "browserstack",
  "bytebase",
  "c",
  "caddy",
  "celery",
  "checkov",
  "chef",
  "circleci",
  "cirrusci",
  "clickhouse",
  "clicksend",
  "cloudbees",
  "cloudbees.codeship",
  "cloudflare",
  "cloudflare.pages",
  "cloudflare.workers",
  "cockroachdb",
  "codecov",
  "codesandboxci",
  "commercetools",
  "consul",
  "contenful",
  "couchbase",
  "coveralls",
  "cplusplus",
  "cratedb",
  "crowdin",
  "csharp",
  "css",
  "cypressci",
  "d3js",
  "dart",
  "databricks",
  "datadog",
  "dataiku",
  "datastax",
  "datocms",
  "deferrun",
  "deno",
  "denodeploy",
  "dependabot",
  "digitalocean",
  "discord",
  "discourse",
  "docker",
  "doctrinephp",
  "docusaurus",
  "docusign",
  "droneci",
  "dropbox",
  "drupal",
  "duckdb",
  "dynatrace",
  "elasticcloud",
  "elasticsearch",
  "elasticstack",
  "electron",
  "eleventy",
  "elixir",
  "emberjs",
  "equinix",
  "esbuild",
  "eslint",
  "expodev",
  "expojs",
  "express",
  "fabric",
  "facebook",
  "fastify",
  "fastly",
  "figma",
  "firebase",
  "firebase.firestore",
  "flyio",
  "flyway",
  "gatsby",
  "gcp",
  "gcp.aiplatform",
  "gcp.appengine",
  "gcp.artifactregistry",
  "gcp.bigquery",
  "gcp.bigtable",
  "gcp.cloudbuild",
  "gcp.cloudrun",
  "gcp.containerregistry",
  "gcp.dataflow",
  "gcp.dataproc",
  "gcp.datastore",
  "gcp.dialogflow",
  "gcp.dns",
  "gcp.functions",
  "gcp.gce",
  "gcp.gcs",
  "gcp.gke",
  "gcp.kms",
  "gcp.language",
  "gcp.logging",
  "gcp.maps",
  "gcp.memorystore",
  "gcp.pubsub",
  "gcp.secretmanager",
  "gcp.spanner",
  "gcp.speech",
  "gcp.sql",
  "gcp.tasks",
  "gcp.translate",
  "gcp.vision",
  "ghost",
  "gitbook",
  "gitguardian",
  "github",
  "github.actions",
  "github.pages",
  "gitlab",
  "gitlab.ci",
  "goacmelego",
  "golang",
  "golangcilint",
  "googleanalytics",
  "grafana",
  "gridsome",
  "haproxy",
  "hashicorp_vault",
  "healthchecksio",
  "helm",
  "heroku",
  "hexojs",
  "hotjar",
  "httpd",
  "hubspot",
  "huggingface",
  "hugo",
  "hyperdx",
  "hypertune",
  "ibmcloud",
  "iftt",
  "influxdb",
  "intercom",
  "intuit",
  "java",
  "jekyll",
  "jenkins",
  "joomla",
  "k6",
  "kentico",
  "kibana",
  "klarna",
  "klaviyo",
  "koa",
  "koalaanalytics",
  "kong",
  "koyeb",
  "kubernetes",
  "lago",
  "laravel",
  "launchdarkly",
  "lemonsqueezy",
  "lighthouse",
  "linear",
  "liquibase",
  "logrocket",
  "logsnag",
  "logstash",
  "magento",
  "mailchimp",
  "mailgun",
  "mailhog",
  "mailjet",
  "mariadb",
  "matomo",
  "mattermost",
  "meilisearch",
  "meilisearchcloud",
  "memcached",
  "meteorjs",
  "milvusdb",
  "mithriljs",
  "mixpanel",
  "mongodb",
  "mongodbatlas",
  "mongoexpress",
  "mysql",
  "nats",
  "neo4j",
  "neondb",
  "netlify",
  "newrelic",
  "nextcloud",
  "nextjs",
  "nginx",
  "nodejs",
  "notion",
  "novu",
  "nuxtjs",
  "oceanbase",
  "okta",
  "onlineornot",
  "openai",
  "optimizely",
  "oraclecloud",
  "ovh",
  "ovh.database",
  "ovh.dedicated",
  "ovh.kubernetes",
  "ovh.storage",
  "ovh.vps",
  "pagerduty",
  "papertrail",
  "payloadcms",
  "paypal",
  "percona",
  "phaserjs",
  "php",
  "phppest",
  "phpstan",
  "phpunit",
  "pingdom",
  "pirschanalytics",
  "placekit",
  "planetscale",
  "platformsh",
  "plausible",
  "playwright",
  "postgresql",
  "posthog",
  "postman",
  "postmark",
  "preactjs",
  "prestashop",
  "prettier",
  "prisma",
  "prismacloud",
  "prometheus",
  "puppeteer",
  "python",
  "qdrant",
  "qovery",
  "qovery.cluster",
  "qovery.database",
  "questdb",
  "rabbitmq",
  "railway",
  "railway.mongodb",
  "railway.mysql",
  "railway.postgres",
  "railway.redis",
  "react",
  "reactemail",
  "redis",
  "redwoodjs",
  "refinedev",
  "relativeci",
  "remixrun",
  "render",
  "renovate",
  "replit",
  "replit.database",
  "replit.postgres",
  "resend",
  "rethinkdb",
  "rollbar",
  "rollup",
  "ruby",
  "rust",
  "salesforce",
  "sanity",
  "sap",
  "sas",
  "scaleway",
  "scaleway.container",
  "scaleway.database",
  "scaleway.documentdb",
  "scaleway.elasticmetal",
  "scaleway.function",
  "scaleway.kubernetes",
  "scaleway.mq",
  "scaleway.redis",
  "scaleway.secretmanager",
  "scaleway.storage",
  "scoutapm",
  "scss",
  "selenium",
  "sendgrid",
  "sentry",
  "sequelize",
  "servicenow",
  "shopify",
  "shopify.hydrogen",
  "signoz",
  "sitecore",
  "sitecore.xmlcloud",
  "slack",
  "snowflake",
  "snyk",
  "socketio",
  "solidjs",
  "sonarcloud",
  "sonarlint",
  "sonarqube",
  "splitio",
  "splunk",
  "sqlite",
  "sqreen",
  "squarespace",
  "squareup",
  "storybook",
  "strapi",
  "stripe",
  "styleci",
  "stylelint",
  "supabase",
  "supabase.functions",
  "supabase.postgres",
  "supabase.realtime",
  "supabase.storage",
  "surrealdb",
  "sveltejs",
  "symfony",
  "tailwind",
  "tdengine",
  "teamcity",
  "teamspeak",
  "telegraf",
  "telegram",
  "tencentcloud",
  "tensorflow",
  "terraform",
  "terragrunt",
  "tidb",
  "timescaledb",
  "tinybird",
  "traefik",
  "travisci",
  "trufflesecurity",
  "twigphp",
  "twilio",
  "twitter",
  "typescript",
  "typesense",
  "typesensecloud",
  "upstash",
  "upstash.kafka",
  "upstash.qstash",
  "upstash.redis",
  "uptimekuma",
  "vale",
  "vercel",
  "vercel.ai",
  "vercel.analytics",
  "vercel.blob",
  "vercel.edge",
  "vercel.kv",
  "vercel.postgres",
  "veriff",
  "vialink",
  "victoriametrics",
  "vite",
  "vue",
  "webflow",
  "webpack",
  "wiz",
  "woocommerce",
  "wordpress",
  "yii2",
  "yousign",
  "zapier",
  "zendesk",
  "zig",
  "zipkin",
  "zookeeper",
  "zoom",
  "zuora",
]
`;
